### DISCUSSION TITLE
MTS200 Thermostat

### DISCUSSION THREAD

**Discussion by chrisburrows** [krahabb/meross_lan#233] [Q&A] [answered]
Firstly let me say a big thank you for this work.... it's a huge step towards getting rid of IoT cloud dependencies.

I have several power plugs connected to HA via the HTTP interface (they're all HomeKit so I've been using that integration). I now have an MTS200 thermostat which also works via the HTTP integration but still uses cloud MQTT.

If the thermostat is switched to local MQTT, is it possible to connect the Meross App to this MQTT broker? The HTTP (and HomeKit) interface doesn't seem to support setting schedules just temperature and mode, so the Meross App is required to manage the schedules. Alternatively are the MQTT messages for configuring heating schedules documented so a UI could be created in HA?

I know I could use HA to completely control the heating, but I still want the thermostat to operate independently and just use HA to add home / away intelligence.

Thanks
Chris

**Comment by krahabb** [comment, marked as answer]
Hello @chrisburrows,
Thank you for your comments! 
Getting the app to connect to a local broker would be awesome but it's not something easy I guess. Since the broker address(es) is not configurable in app the only way I see would be to 'fake-route-proxy' the connection to the official meross cloud servers by using a lan router able to redirect traffic (or just re-mapping to a local broker the dns queries for the offcial servers). At any rate this is only the first step since there are some rules in place on the official servers in order to let the app communicate with the devices (think about all the authentication and access check stuff) so that's really not easy (should be doable though). You can check @albertogeniola wikis about the MQTT message flow on the remote servers.

The other topic (managing the device own schedule) would be easier and the configuration messages are easy to understand. It's just it would need a lot of UI and I never spent time on this.
The message to query and set the MTS200 schedule is 'Appliance.Control.Thermostat.Schedule' and it should work both on MQTT and HTTP. I remember I spent some time inspecting the message payload and it was really straightforward but, again, it would need a bit of effort on the UI side to be able to expose a nice user experience.
I'm up to help you on this, expecially if you come around with some nice ideas on how this UI should behave but I don't have enough time now to implement a 'full featured' experience.

An idea, to start with, would be to implement a service in meross_lan to ease the setting of the schedule parameters so you could use scripts and automations to interact your own way. There's actually a very basic service call available which can be used to send raw payloads to the device but of course there's a lot of effort needed to build the correct payload syntax for the specific message

**Reply by chrisburrows** [reply]
Thank you for your reply. I had assumed that doing anything with the Meross App would be almost impossible due to TLS certificate validation.
For the time being I'll leave by thermostat connected to the Meross MQTT, but will have a play with the API via HTTP. Time permitting, I'll start having a look at how a schedule could be created and pushed to the thermostat. Would be great if something could be created on top of Local Calendar, or Nils Faber's scheduler card.

**Comment by krahabb** [comment]
Hello @chrisburrows,
I'm actually testing and developing this feature for the MTS100 line of valves. So far I'm able to show the schedule on a dedicated HA calendar entity and I'm implementing the feature to effectively edit this schedule.

The MTS100 line of products exposes the schedule this way:
```json
    "payload":{
      "schedule": [
        {
          "id": "0100783A",
          "mon": [[390,150],[90,240],[300,190],[270,220],[390,150]],
          "tue": [[390,150],[90,240],[300,190],[270,220],[390,150]],
          "wed": [[390,150],[90,240],[300,190],[270,220],[390,150]],
          "thu": [[390,150],[90,240],[300,190],[270,220],[390,150]],
          "fri": [[390,150],[90,240],[300,190],[270,220],[390,150]],
          "sat": [[390,150],[90,240],[300,190],[270,220],[390,150]],
          "sun": [[390,150],[90,240],[300,190],[270,220],[390,150]]
        }
      ]
    }
```
This shows the schedule must be set on 6 intervals each day and repeats weekly: every interval here represents a duration in minutes and the setpoint temperature (in tenths of °C)
In my experiments I was able to edit this schedule with the only limit the intervals must be 6 for each day (I'm pretty sure on the app you can change this but I didn't find a way so far)

If you can 'grab' a trace of your device (MTS200) schedule I could start mock-up an implementation for this device too: it shouldnt be that different
